specs/ClientCertificate.md

```diff
@@ -0,0 +1,380 @@
+## Win32 C++
+```cpp
+// Turn on or off client certificate selection dialog by adding or removing a
+// ClientCertificateRequested handler which intercepts when server requests client
+// certificate for authentication.
+void SettingsComponent::SetClientCertificateRequested(bool raiseClientCertificate)
+{
+    if (raiseClientCertificate != m_raiseClientCertificate)
+    {
+        m_raiseClientCertificate = raiseClientCertificate;
+        //! [ClientCertificateRequested]
+        if (m_raiseClientCertificate)
+        {
+            CHECK_FAILURE(m_webView->add_ClientCertificateRequested(
+                Callback<ICoreWebView2ClientCertificateRequestedEventHandler>(
+                    [this](
+                        ICoreWebView2* sender,
+                        ICoreWebView2ClientCertificateRequestedEventArgs* args) {
+
+                        ICoreWebView2ClientCertificateList* certificateList;
+                        CHECK_FAILURE(args->get_MutuallyTrustedCertificates(&certificateList));
+
+                        wil::unique_cotaskmem_string host;
+                        CHECK_FAILURE(args->get_Host(&host));
+
+                        BOOL isProxy = FALSE;
+                        CHECK_FAILURE(args->get_IsProxy(&isProxy));
+
+                        UINT cert_lst_size;
+                        CHECK_FAILURE(certificateList->get_Count(&cert_lst_size));
+
+                        ICoreWebView2ClientCertificate* certificate = nullptr;
+                        if (cert_lst_size > 0)
+                        {
+                            for (int i = 0; i < cert_lst_size; i++)
+                            {
+                                CHECK_FAILURE(certificateList->GetValueAtIndex(i, &certificate));
+
+                                wil::unique_cotaskmem_string subject;
+                                CHECK_FAILURE(certificate->get_Subject(&subject));
+                                wil::unique_cotaskmem_string issuer;
+                                CHECK_FAILURE(certificate->get_Issuer(&issuer));
+                                wil::unique_cotaskmem_string serialNumber;
+                                CHECK_FAILURE(certificate->get_SerialNumber(&serialNumber));
+                                COREWEBVIEW2_CLIENT_CERTIFICATE_KIND certificateKind;
+                                CHECK_FAILURE(certificate->get_CertificateKind(&certificateKind));
+                                INT64 startDate;
+                                CHECK_FAILURE(certificate->get_ValidStartDate(&startDate));
+                                INT64 expiryDate;
+                                CHECK_FAILURE(certificate->get_ValidExpiryDate(&expiryDate));
+                                ICoreWebView2ClientCertificateIssuerChainList* pemIssuerChainList;
+                                CHECK_FAILURE(certificate->get_PEMEncodedIssuerChainData(&pemIssuerChainList));
```

Comment:
This sample is a little weird. We get a bunch of properties but don't do anything with them. And then after the loop we check `if (certificate)` which will contain just the last certificate we looked at in the for loop?

Maybe you intended to gather up all these properties to initialize a custom UI? And then just after the for loop but before the `if (certificate)` check you would show the UI? If we're showing UI we should do it asynchronoulsy, by taking a deferral and then queuing a lambda to run not during the event handler.

Reply:
We can use the API mainly for two purposes, to set the client certificate selection programmatically without showing dialog or to show custom dialog instead of default one. 

My previous example is not clear enough like you said. Added two samples for win32 app, example 1 sets the certificate without showing dialog to the end user and example 2 displays a dialog with certificate list for selection instead of default one by taking deferral